src/vmware/vmxnet3_net.c

```diff
@@ -0,0 +1,581 @@
+
+    vmx_ds->vmxnet3_revision = VMXNET3_REVISION;
+    vmx_ds->upt_version = VMXNET3_UPT_VERSION;
+    vmx_ds->upt_features = UPT1_F_CSUM | UPT1_F_LRO;
+    vmx_ds->driver_data = physical_from_virtual(dev);
+    vmx_ds->driver_data_len = sizeof(struct vmxnet3);
+    // queue_shared & queue_shared_len are in
+    // vmxnet3_queues_shared_alloc()
+    vmx_ds->mtu = 1500;
+    vmx_ds->nrxsg_max = VMXNET3_MAX_RX_SEGS;
+    vmx_ds->ntxqueue = VMXNET3_DEF_TX_QUEUES;
+    vmx_ds->nrxqueue = VMXNET3_DEF_RX_QUEUES;
+
+    vmx_ds->automask = 0;
+    vmx_ds->nintr = 2;
+    vmx_ds->evintr = 0;
+
+    vmx_ds->rxmode = VMXNET3_RXMODE_UCAST | VMXNET3_RXMODE_MCAST | VMXNET3_RXMODE_BCAST |
+        VMXNET3_RXMODE_ALLMULTI;
+
+    vmxnet3_alloc_mcast_table(dev);
+
+    u64 ph = physical_from_virtual(vmx_ds);
+    pci_bar_write_4(&dev->bar1, VMXNET3_BAR1_DSL, ph);
+    pci_bar_write_4(&dev->bar1, VMXNET3_BAR1_DSH, ph >> 32);
+
+    vmxnet3_write_cmd(dev, VMXNET3_CMD_SET_FILTER);
+    vmxnet3_write_cmd(dev, VMXNET3_CMD_SET_RXMODE);
+}
+
+static void
+vmxnet3_interrupts_disable(vmxnet3_pci dev)
+{
+    struct vmxnet3_rxqueue *rxq = dev->vmx_rxq[0];
+    struct vmxnet3_txqueue *txq = dev->vmx_txq[0];
+    pci_bar_write_4(&dev->bar0, VMXNET3_BAR0_IMASK(rxq->vxrxq_intr_idx), 1);
+    pci_bar_write_4(&dev->bar0, VMXNET3_BAR0_IMASK(txq->vxtxq_intr_idx), 1);
+}
+
+static void
+kick_pending(vmxnet3_pci dev)
+{
+    struct vmxnet3_txqueue* vmx_txq = dev->vmx_txq[0];
+    if (vmx_txq->vxtxq_ts->npending) {
+        vmx_txq->vxtxq_ts->npending = 0;
+        pci_bar_write_4(&dev->bar0, VMXNET3_BAR0_TXH(0), vmx_txq->vxtxq_cmd_ring.vxtxr_head);
+    }
+}
+
+static err_t
+low_level_output(struct netif *netif, struct pbuf *p)
+{
+    vmxnet3 vn = netif->state;
+
+    err_t e = vmxnet3_isc_txd_encap(vn->dev, p);
+    if (e != ERR_OK)
+        return e;
+    kick_pending(vn->dev)
+
+    MIB2_STATS_NETIF_ADD(netif, ifoutoctets, p->tot_len);
+    if (((u8_t *)p->payload)[0] & 1) {
+        /* broadcast or multicast packet*/
+        MIB2_STATS_NETIF_INC(netif, ifoutnucastpkts);
+    } else {
+        /* unicast packet */
+        MIB2_STATS_NETIF_INC(netif, ifoutucastpkts);
+    }
+    /* increase ifoutdiscards or ifouterrors on error */
+
+    LINK_STATS_INC(link.xmit);
+
+    return ERR_OK;
+}
+
+static void
+vmxnet3_get_mac(vmxnet3 vx)
+{
+    u32 ml = vmxnet3_read_cmd(vx->dev, VMXNET3_CMD_GET_MACL);
+    u32 mh = vmxnet3_read_cmd(vx->dev, VMXNET3_CMD_GET_MACH);
+
+    struct netif* netif = vx->n;
+
+    netif->hwaddr[0] = ml;
+    netif->hwaddr[1] = ml >> 8;
+    netif->hwaddr[2] = ml >> 16;
+    netif->hwaddr[3] = ml >> 24;
+    netif->hwaddr[4] = mh;
+    netif->hwaddr[5] = mh >> 8;
+}
+
+static void
+vmxnet3_check_version(vmxnet3_pci dev)
+{
+    u32 version = pci_bar_read_4(&dev->bar1, VMXNET3_BAR1_VRRS);
+    if ((version & 0x01) == 0) {
+        rprintf("unsupported hardware version %x\n", version);
+        assert(false);
```

Comment:
halt(msg, ...) may be a little neater here (and below) - avoids an unnecessary stack dump